semimpl/do_test.go

```diff
@@ -0,0 +1,181 @@
+package semimpl_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/benchttp/runner/semimpl"
+)
+
+func TestDo(t *testing.T) {
+	t.Run("stop until maxIter is reached", func(t *testing.T) {
+		var (
+			numWorkers = 1
+			gotIter    = 0
+			maxIter    = 10
+			expIter    = 10
+		)
+
+		semimpl.Do(context.Background(), numWorkers, maxIter, func() {
+			gotIter++
+		})
+
+		if gotIter != expIter {
+			t.Errorf("iterations: exp %d, got %d", expIter, gotIter)
+		}
+	})
+
+	t.Run("stop on context timeout", func(t *testing.T) {
+		var (
+			timeout    = 50 * time.Millisecond
+			interval   = timeout / 5
+			numWorkers = 1
+			maxIter    = 0 // infinite iterations
+
+			margin      = 15 * time.Millisecond
+			maxDuration = timeout + margin
+		)
+
+		ctx, cancel := context.WithTimeout(context.Background(), timeout)
+		defer cancel()
+
+		gotDuration := timeFunc(func() {
+			semimpl.Do(ctx, numWorkers, maxIter, func() {
+				time.Sleep(interval)
+			})
+		})
+
+		if gotDuration > maxDuration {
+			t.Errorf(
+				"context timeout duration: exp < %dms, got %dms",
+				maxDuration.Milliseconds(), gotDuration.Milliseconds(),
+			)
+		}
+	})
+
+	t.Run("stop on context cancel", func(t *testing.T) {
+		var (
+			timeout    = 50 * time.Millisecond
+			interval   = 10 * time.Millisecond
+			numWorkers = 1
+			maxIter    = 0 // infinite iterations
+
+			margin      = 15 * time.Millisecond
+			maxDuration = timeout + margin
+		)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			time.Sleep(timeout)
+			cancel()
+		}()
+
+		gotDuration := timeFunc(func() {
+			semimpl.Do(ctx, numWorkers, maxIter, func() {
+				time.Sleep(interval)
+			})
+		})
+
+		if gotDuration > maxDuration {
+			t.Errorf(
+				"context cancel duration: exp < %dms, got %dms",
+				maxDuration.Milliseconds(), gotDuration.Milliseconds(),
+			)
+		}
+	})
+
+	t.Run("limit concurrent workers", func(t *testing.T) {
+		var (
+			numWorkers = 3
+			maxIter    = 12
+
+			durations                = make([]time.Duration, 0, maxIter)
+			maxIntervalWithinGroup   = 10 * time.Millisecond
+			minIntervalBetweenGroups = 30 * time.Millisecond
+		)
+
+		start := time.Now()
+		semimpl.Do(context.Background(), numWorkers, maxIter, func() {
+			durations = append(durations, time.Since(start))
```

Comment:
Note: this assignation is technically not thread-safe. However with only concurrent 3 workers the odds of a race condition is probably negligible
If that bothers someone I'll update it.

Reply:
It is very unlikely there is a problem, but I feel like it would be cleaner to make it thread-safe.
I guess it could also be considered to be overkill though 🤷‍♀️